icicle_v3/backend/cpu/include/tasks_manager.h

```diff
@@ -0,0 +1,309 @@
+
+    /**
+     * @brief Check if all worker's tasks are idle (free with no result to be handled)
+     */
+    bool are_all_idle();
+  private:
+    std::thread task_executor; // Thread to be run parallel to main
+    std::vector<Task> mTasksFifo; // vector containing the worker's task. a Vector is used to allow buffering.
+    int tail; // Tail (input) idx of the fifo above. Checks for free task start at this idx.
+    int head; // Head (output) idx of the fifo above, the thread works on task at this idx.
+    bool kill; // boolean to flag from main to the thread to finish.
+  };
+
+  std::vector<Worker> workers; // Vector of workers/thread to be ran simultaneously.
+};
+
+/**
+ * @class TaskBase
+ * @brief abstract base for the task supported by TasksManager above.
+ */
+class TaskBase {
+public:
+  /**
+   * @brief constructor fo `TaskBase`.
+   */
+  TaskBase() : status(IDLE), father_fifo_tail(nullptr) {}
+
+  /**
+   * @brief function to be executed by `TasksManager`. Implemented by child class.
+   */
+  virtual void execute() = 0;
+
+  /**
+   * @brief Signal for the `Worker` owning the task that it is ready to be executed. 
+   * Also increase's the workers tail idx.
+   * User function.
+   */
+  void dispatch();
+
+  // Getters and setter for the various states of the tasks `status`.
+  inline bool is_ready_for_work(); // DISPATCHED
+  inline bool can_push_task(); // IDLE or PENDING_RESULT
+  inline bool has_result();
+  inline bool is_idle();
+  inline void set_working();
+  inline void set_pending_result();
+  inline void set_handled_result();
+
+  /**
+   * @brief link this task with the tail idx of the Worker it belongs to. Used only once after construction.
+   */
+  inline void link_father_tail(int* tail_pointer) { father_fifo_tail = tail_pointer; } 
+
+  /**
+   * @brief wait for a specific task to finish executing. This is a blocking function.
+   */
+  void wait_done(); // NOTE the user must handle this task's result before asking for new free tasks  
+
+private:
+  /**
+   * @enum containing the valid states of a task.
+   */
+  enum eTaskStatus {IDLE, DISPATCHED, WORKING, PENDING_RESULT}; // TODO CAPS and eTaskStatus
+  std::atomic<eTaskStatus> status; // current task state. Atomic to ensure proper sync between main and worker threads.
+  int* father_fifo_tail; // pointer to father's tail.
+};
+
+
+bool TaskBase::is_ready_for_work() {
+  return status.load(std::memory_order_acquire) == DISPATCHED;
+}
+
+bool TaskBase::can_push_task() {
+  eTaskStatus curr_status = status.load(std::memory_order_acquire);
+  return curr_status == PENDING_RESULT || curr_status == IDLE;
+}
+
+bool TaskBase::has_result() {
+  return status.load(std::memory_order_acquire) == PENDING_RESULT;
+}
+
+bool TaskBase::is_idle() {
+  return status.load(std::memory_order_acquire) == IDLE;
+}
+
+void TaskBase::dispatch() {
+  (*father_fifo_tail)++;
+  status.store(DISPATCHED, std::memory_order_release);
+}
+
+void TaskBase::set_working() {
```

Comment:
don't you check valid state transitions? Don't assume only happy path

Reply:
I had asserts but removed them
Should I re-add them? Or throw a sort of Icicle error instead?

(all the status functions should only be accessed by the worker anyway - maybe changing to private + friend would be better)